src/connection_manager.c

```diff
@@ -438,25 +438,37 @@ static qd_error_t load_server_config(qd_dispatch_t *qd, qd_server_config_t *conf
         config->max_frame_size = QD_AMQP_MIN_MAX_FRAME_SIZE;
 
     //
-    // Given session frame count and max frame size compute session incoming_capacity
+    // Given session frame count and max frame size, compute session incoming_capacity
+    //   On 64-bit systems the capacity has no limit.
+    //   On 32-bit systems the largest capacity is AMQP_MAX_WINDWOW_SIZE.
     //
-    if (ssn_frames == 0)
-        config->incoming_capacity = (sizeof(size_t) < 8) ? 0x7FFFFFFFLL : 0x7FFFFFFFLL * config->max_frame_size;
-    else {
-        uint64_t mfs      = (uint64_t) config->max_frame_size;
-        uint64_t trial_ic = ssn_frames * mfs;
-        uint64_t limit    = (sizeof(size_t) < 8) ? (1ll << 31) - 1 : 0;
-        if (limit == 0 || trial_ic < limit) {
-            // Silently promote incoming capacity of zero to one
-            config->incoming_capacity = 
-                (trial_ic < QD_AMQP_MIN_MAX_FRAME_SIZE ? QD_AMQP_MIN_MAX_FRAME_SIZE : trial_ic);
+    if (ssn_frames == 0) {
+        // Unlimited incoming frames.
+        // config->incoming_capacity is zero. Proton incoming_window is always AMQP_MAX_WINDOW_SIZE
+    } else {
+        // Limited incoming frames.
+        // Specify this to proton by setting capacity to be
+        // the product (max_frame_size * ssn_frames).
+
+        size_t capacity = config->max_frame_size * ssn_frames;
+        assert(capacity >= (size_t)QD_AMQP_MIN_MAX_FRAME_SIZE);
```

Comment:
The assert seems redundant if there is any verification of the actual max frame size setting..and if there isnt, maybe there should be? The ssn_frames is known to be non zero, and if its >=1 the assert cant trigger unless the frame size is illegal.

Ah, unless the assert  is also trying to handle overflow in the capacity calculation? In which case it doesnt really do that very completely, just checks it isnt under max frame size.